grid/studies/seamstress/index.md

```diff
@@ -0,0 +1,777 @@
+-- without an argument, seamstress will always connect to the first-registered grid.
+
+function init()
+  if g then
+    cols = g.cols
+    rows = g.rows
+    grid_connected = true
+  else
+    cols = 16
+    rows = 8
+    grid_connected = false
+  end
+
+  playhead = clock.run(play)
+  grid_dirty = true
+  grid_redraw = metro.init(
+    draw_grid, -- function to execute
+    1 / 60, -- how often (here, 60 fps)
+    -1 -- how many times (here, forever)
+  )
+  grid_redraw:start() -- start the timer
+  
+end
+
+function grid.add(dev)
+  cols = dev.cols
+  rows = dev.rows
+  grid_connected = true
+end
+
+function grid.remove(dev)
+  grid_connected = false
+end
+
+function play()
+  while true do
+    -- perform actions
+    clock.sync(1 / 4)
+    grid_dirty = true
+  end
+end
+
+function g.key(x, y, z)
+  -- define grid keypress action
+  grid_dirty = true
+end
+
+function draw_grid()
+  if grid_dirty then
+    g:all(0) -- clear grid LEDs
+    -- queue grid LEDs
+    g:refresh() -- draw grid LEDs
+    grid_dirty = false -- reset flag
+  end
+end
+```
+
+`grid.add` and `grid.remove` are two callback functions built into seamstress, which respond when a grid is physically connected or disconnected to the host computer. The `dev` table each function returns holds the following information:
+
+- `cols` and `rows` for the grid
+- `id` of the grid (also printed at script boot)
+- `port` the grid is connected to
+- `name` and `serial` of the grid
+
+So we can write a responsive interface for our step sequencer, we establish a few variables:
+
+- `cols`, which we'll use to set the maximum number of steps
+- `rows`, which we'll to set the maximum number of tracks
+- `grid_connected`, which we'll use to query whether there's a grid to communicate with
+
+### Schedule, Process, Display {#schedule-process-display}
+
+In our script's `init`, we schedule two functions to be executed at regular intervals.
+
+#### clock
+
+The first is done with the `clock` library, which allows tempo-synced scheduling:
+
+```lua
+playhead = clock.run(play)
+
+[...]
+
+function play()
+  while true do
+    -- perform actions
+    grid_dirty = true
+    clock.sync(1/4)
+  end
+end
+```
+
+In `play`, we perform our actions and then sync to the clock at a quarter pulse (which is a 1/16th note in 4/4).
+
+#### metro
+
+The second is done with the `metro` library, which allows scheduling with high-accuracy timers:
+
+```lua
+grid_redraw = metro.init(
+  draw_grid, -- function to execute
+  1/60, -- how often (here, 60 fps)
+  -1 -- how many times (here, forever)
+)
+grid_redraw:start() -- start the timer
+[...]
+
+function draw_grid()
+  if grid_dirty then
+    g:all(0) -- clear grid LEDs
+    -- queue grid LEDs
+    g:refresh() -- draw grid LEDs
+    grid_dirty = false -- reset flag
+  end
+end
+```
+
+Since we want the grid to draw at a steady 60 frames-per-second, we opt for the higher-resolution `metro`, rather than the 'musical timing' `clock`.
```

Comment:
I was gonna complain that clocks and metros are really the same mechanism, but I suppose they're *not* anymore, so this is a fair statement :)